meta-client/src/main/java/org/apache/helix/metaclient/api/MetaClientInterface.java

```diff
@@ -20,139 +20,382 @@
+   * @param key For metadata storage that has hierarchical key space (e.g. ZK), the key would be
+   *            a parent key,
+   *            For metadata storage that has non-hierarchical key space (e.g. etcd), the key would
+   *            be a prefix key.
+   * @eturn Return a list of sub entry keys. Return direct child name only for hierarchical key
+   *        space, return the whole sub key for non-hierarchical key space.
    */
-  List<String> getSubEntryKeys(final String path);
+  List<String> getSubEntryKeys(final String key);
 
   /**
-   * Return the number of sub entries for the given keys
-   * @param path: For metadata storage that has hierarchical key space (e.g. ZK), the path would be
-   *            a parent path,
-   *            For metadata storage that has non-hierarchical key space (e.g. etcd), the path would
-   *            be a prefix path.
+   * Return the number of sub entries for the given keys.
+   * @param key For metadata storage that has hierarchical key space (e.g. ZK), the key would be
+   *            a parent key,
+   *            For metadata storage that has non-hierarchical key space (e.g. etcd), the key would
+   *            be a prefix key.
    */
-  int countSubEntries(final String path);
+  int countSubEntries(final String key);
 
   /**
    * Remove the entry associated with the given key.
    * For metadata storage that has hierarchical key space, the entry can only be deleted if the key
    * has no child entry.
-   * TODO: throws
-   * @param path
-   * @return
+   * TODO: define exception to throw
+   * @param key  key to identify the entry to delete
+   * @return Return true if the deletion is completed
    */
-  boolean delete(String path);
+  boolean delete(final String key);
 
   /**
    * Remove the entry associated with the given key.
    * For metadata storage that has hierarchical key space, remove all its child entries as well
    * For metadata storage that has non-hierarchical key space, this API is the same as delete()
-   * @param path
-   * @return
+   * @param key key to identify the entry to delete
+   * @return Return true if the deletion is completed
    */
-  boolean recursiveDelete(String path);
+  boolean recursiveDelete(final String key);
 
   /* Asynchronous methods return immediately.
    * They take a callback object that will be executed either on successful execution of the request
    * or on error with an appropriate return code indicating the error.
    */
-  void asyncCreate(final String key, T data, int version, long ttl,
+
+  /**
+   * The asynchronous version of create.
+   * @param key key to identify the entry
+   * @param data value of the entry
+   * @param mode EntryMode identifying if the entry will be deleted upon client disconnect
+   * @param cb An user defined VoidCallback implementation that will be invoked when async create return.
+   */
+  void asyncCreate(final String key, final T data, final EntryMode mode,
```

Comment:
All these async APIs, shall we return some results? Or we rely on callbacks? How about exceptions happen?

Reply:
We rely on callbacks. The return code in callback will include exception type. We remain the same behavior as ZK.
And will define the exceptions in following commits.